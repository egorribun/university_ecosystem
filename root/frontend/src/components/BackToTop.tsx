import { useEffect, useState, useCallback } from "react";

export default function BackToTop() {
  const [show, setShow] = useState(false);
  useEffect(() => {
    const onScroll = () => setShow(window.scrollY > 420);
    onScroll();
    window.addEventListener("scroll", onScroll, { passive: true });
    return () => window.removeEventListener("scroll", onScroll);
  }, []);
  const onClick = useCallback(() => {
    try { window.scrollTo({ top: 0, behavior: "smooth" }); } catch { window.scrollTo(0, 0); }
  }, []);
  return (
    <button
      type="button"
      className={"back-to-top" + (show ? " visible" : "")}
      aria-label="Наверх"
      onClick={onClick}
    >
      ↑
    </button>
  );
}