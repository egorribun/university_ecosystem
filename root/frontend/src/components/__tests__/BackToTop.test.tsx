import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import BackToTop from "../BackToTop";

const setScrollY = (value: number) => {
  Object.defineProperty(window, "scrollY", { value, configurable: true });
  Object.defineProperty(window, "pageYOffset", { value, configurable: true });
};

describe("BackToTop", () => {
  beforeEach(() => {
    setScrollY(0);
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("renders hidden button by default", () => {
    const { container } = render(<BackToTop />);
    const button = screen.getByRole("button", { name: "Наверх" });
    expect(button).toBeInTheDocument();
    expect(button).not.toHaveClass("visible");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("toggles visibility after scrolling", async () => {
    render(<BackToTop />);
    const button = screen.getByRole("button", { name: "Наверх" });

    setScrollY(500);
    fireEvent.scroll(window);

    await waitFor(() => expect(button).toHaveClass("visible"));

    setScrollY(0);
    fireEvent.scroll(window);

    await waitFor(() => expect(button).not.toHaveClass("visible"));
  });

  it("scrolls smoothly to top when clicked", () => {
    const scrollTo = vi.spyOn(window, "scrollTo").mockImplementation(() => {});
    const { getByRole } = render(<BackToTop />);
    const button = getByRole("button", { name: "Наверх" });

    setScrollY(500);
    fireEvent.scroll(window);

    fireEvent.click(button);

    expect(scrollTo).toHaveBeenCalledTimes(1);
    expect(scrollTo).toHaveBeenCalledWith({ top: 0, behavior: "smooth" });
  });
});
