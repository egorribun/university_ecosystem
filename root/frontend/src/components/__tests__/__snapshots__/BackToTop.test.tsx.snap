// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackToTop > renders hidden button by default 1`] = `
<button
  aria-label="Наверх"
  class="back-to-top"
  type="button"
>
  ↑
</button>
`;
